src/mpid/ch4/src/ch4_impl.h

```diff
@@ -878,6 +879,33 @@ MPL_STATIC_INLINE_PREFIX void *MPIDIU_map_update(void *in_map, uint64_t id, void
     return rc;
 }
 
+/* Return the associated av for a RMA target and the win attributes embedded into win handle.
+ * This is an optimized path for direct intra comm (comm_world or dup from comm_world) by
+ * eliminating pointer dereferences into dynamic allocated objects (i.e., win->comm_ptr).*/
+MPL_STATIC_INLINE_PREFIX MPIDI_av_entry_t *MPIDIU_win_rank_to_av(MPIR_Win * win, int rank,
+                                                                 MPIDI_winattr_t * winattr)
+{
+    MPIDI_av_entry_t *av = NULL;
+
+    if (HANDLE_CHECK_MPI_RESERVE_BIT(win->handle)) {
```

Comment:
why a handle bit and not a field in the win struct?

Reply:
The intention was to save pointer dereferences into the window object. In my prototype code for OSHMPI analysis, I used directly the input handle at MPI_Put. But then we had to break the MPID API in order to pass down the `winattr` parameter.

This PR used `win->handle` at MPID_put in order to avoid breaking MPID API, which does cost a pointer deference. I guess it is the same as using a field of win now.

Do you have any better idea (i.e., how to pass down handle encoded info without breaking MPID API)?

Reply:
> The intention was to save pointer dereferences into the window object. In my prototype code for OSHMPI analysis, I used directly the input handle at MPI_Put. But then we had to break the MPID API in order to pass down the `winattr` parameter.
> 
> This PR used `win->handle` at MPID_put in order to avoid breaking MPID API, which does cost a pointer deference. I guess it is the same as using a field of win now.
> 
> Do you have any better idea (i.e., how to pass down handle encoded info without breaking MPID API)?

Ah, yes. For datatypes, we pass the handle all the way down and do not dereference it in the common case. Unfortunately I don't see a way to do this with window handles without a change in the ADI.

Reply:
Is it possible to change ADI? e.g., after the upcoming big release?

Reply:
> Is it possible to change ADI? e.g., after the upcoming big release?

I think it would be worth considering. IMO, it's not too difficult to adapt to using handle vs. ptr in this case.

Reply:
At MPI_Put, we already get the win ptr for error checking. Getting the ptr costs 9 instructions. The best would be pass down both ptr and handle, if it is not too messy.

Reply:
I'll have to think about it. Converting from handle->ptr is not expensive, but if it needs to be done in multiple layers of the call stack, that would not be ideal. My reaction to passing both is it sounds a bit clunky.

Reply:
Agreed with both concerns. For this PR, I would just get rid of the handle embedding part, and use a flag field of win object instead. It still avoids more expensive win->comm_ptr dereferences.

But for the next step, I would think passing down the handle itself might be worth to consider, at least for the functions that are at fast-path.

Reply:
Sounds good to me.

Reply:
> At MPI_Put, we already get the win ptr for error checking. Getting the ptr costs 9 instructions. The best would be pass down both ptr and handle, if it is not too messy.

Error checking can be configured off, right? I don't think we should make an optimization decision based on an optional part. I mean, without error checking, does the argument -- duplicated cost of getting the ptr -- still hold?

Reply:
@hzhou raised a good point. But error checking is enabled by default. I think we still want to avoid getting the ptr twice in such a case. It just looks bad from engineer's view.

If we move the error checking down to `MPID_` layer, we don't need get ptr any more at `MPI_` layer. But I guess that will force duplicated code in both ch3 and ch4, which is not good too.

I am afraid that the only way to have handle-based optimization is to add a parameter (e.g., `MPI_Win win_handle, MPIR_Win *win`) in the `MPID` API.

Reply:
From "manager" point of view -- 😄 -- if the application can afford turning on error checking, then it should afford the extra 9 instructions. A scenario could be for the application to run with error checking on to ensure correctness, then run with error checking off for the bulk task. The extra parameter in the ADI for specific optimization is obscure, very difficult to comprehend for new developers, so I am leaning toward not to complicate the interface unless the gain is justifiable, such as real measurable benefit for real applications.

Reply:
We can discuss further in a separate PR. This PR will just add a `winattr` field in the win object.